hypothesis-python/src/hypothesis/strategies/_internal/numbers.py

```diff
@@ -195,24 +202,42 @@ def __init__(self, allow_infinity, allow_nan, allow_subnormal, width):
         self.allow_nan = allow_nan
         self.allow_subnormal = allow_subnormal
         self.width = width
-
-        self.nasty_floats = [
-            float_of(f, self.width) for f in NASTY_FLOATS if self.permitted(f)
-        ]
-        weights = [0.2 * len(self.nasty_floats)] + [0.8] * len(self.nasty_floats)
-        self.sampler = d.Sampler(weights)
+        self.bounds = bounds
+        if bounds:
+            (lower_bound, upper_bound) = bounds
+            assert isinstance(lower_bound, float)
+            assert isinstance(upper_bound, float)
+            assert 0 <= lower_bound < upper_bound
+            assert math.copysign(1, lower_bound) == 1, "lower bound may not be -0.0"
+            self.nasty_floats = None
+            self.sampler = None
+        else:
+            self.nasty_floats = [
+                float_of(f, self.width) for f in NASTY_FLOATS if self.permitted(f)
+            ]
+            weights = [0.2 * len(self.nasty_floats)] + [0.8] * len(self.nasty_floats)
+            self.sampler = d.Sampler(weights)
 
     def __repr__(self):
-        return "{}(allow_infinity={}, allow_nan={}, width={})".format(
-            self.__class__.__name__, self.allow_infinity, self.allow_nan, self.width
+        return "{}(allow_infinity={}, allow_nan={}, width={}, bounds={})".format(
+            self.__class__.__name__,
+            self.allow_infinity,
+            self.allow_nan,
+            self.width,
+            self.bounds,
         )
 
     def permitted(self, f):
         assert isinstance(f, float)
-        if not self.allow_infinity and math.isinf(f):
-            return False
-        if not self.allow_nan and math.isnan(f):
-            return False
+        if math.isfinite(f):
+            if self.bounds is not None:
+                if not (self.bounds[0] <= f <= self.bounds[1]):
+                    return False
```

Comment:
Another sticking point - I don't have coverage on this line. Previously this extra check was accomplished with a call to `assume()`. (but it feels like the check belongs here now) I might need z3 to help me find the counterexample which might cause the range math to escape the bounds! Will poke at it.

Reply:
I suspect it might be nontrivially unreachable?  Happy to do a deeper dive once I'm out of the franticness of sprints-day-one 😅

Reply:
I am in no rush! I think Z3 confirms your suspicion (at least in Float16; Float32 doesn't complete in a reasonable timeframe):
```
>>> from z3 import *
>>> top=Const("top", Float16())
>>> bottom=Const("bottom", Float16())
>>> frac=Const("frac", Float16())
>>> s=Solver()
>>> s.add(0.0 <= frac)
>>> s.add(frac <= 1.0)
>>> s.add(0.0 <= bottom)
>>> s.add(bottom <= top)
>>> s.check(bottom + ((top - bottom) * frac) < bottom)
unsat
>>> s.check(bottom + ((top - bottom) * frac) > top)
unsat

```